[CodeGen] Allow mixed scalar type constraints for inline asm

dfszabo opened the pull request:
GCC supports code like "asm volatile ("" : "=r" (i) : "0" (f))" where i is integer type and f is floating point type. Currently this code produces an error with Clang. The change allows mixed scalar types between input and output constraints.

RKSimon commented:
Are there existing bug issues associated with this?

dfszabo commented:
> Are there existing bug issues associated with this?

I don't know. Found this issue in our downstream target.

topperc commented on clang/test/CodeGen/inline-asm-fp-to-int.c:
```diff
@@ -0,0 +1,8 @@
+// RUN: %clang_cc1 %s -emit-llvm -o /dev/null
+
+unsigned test(float f)
```
What IR does this generate?

topperc replied:
This test is using -emit-llvm, but the patch is in SelectionDAG. SelectionDAG doesn't run with -emit-llvm

dfszabo replied:
Thanks for the review. You are right, indeed the test was pointless, the problem is with CodeGen. Updated the issue accordingly.

dfszabo commented:
Ping

dfszabo commented:
Ping

arsenm commented on llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:
```diff
@@ -0,0 +1,32 @@
+  %i = alloca i32, align 4
+  store float %f, ptr %f.addr, align 4
+  %0 = load float, ptr %f.addr, align 4
+  %1 = call i32 asm sideeffect "", "=r,0,~{dirflag},~{fpsr},~{flags}"(float %0)
```
Use named values in tests

arsenm commented on llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:
```diff
@@ -0,0 +1,32 @@
+  store i32 %1, ptr %i, align 4
+  %2 = load i32, ptr %i, align 4
+  ret i32 %2
+}
```
Can you also try the same, except with a mixed pointer and int/float? Also some vector cases?

arsenm commented on llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:
```diff
@@ -0,0 +1,72 @@
+  %f.addr = alloca float*, align 4
+  %i = alloca i32, align 4
+  store float* %f, ptr %f.addr, align 4
+  %load_f = load float*, ptr %f.addr, align 4
```
Test needs to be updated to use opaque pointers. Also you don't need all of this intermediate alloca stuff, you can simplify the incoming values and uses

arsenm reviewed (changes requested):
Test needs cleanup

arsenm commented on llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:
```diff
@@ -0,0 +1,72 @@
+;    return i;
+; }
+
+
```
Add some comments explaining what this is showing

dfszabo committed:
[CodeGen] Allow mixed scalar type constraints for inline asm

GCC supports code like "asm volatile ("" : "=r" (i) : "0" (f))" where i is integer type and f is floating point type. Currently this code produces an error with Clang. The change allows mixed scalar types between input and output constraints.

arsenm commented on llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:
```diff
@@ -0,0 +1,61 @@
+; CHECK-NEXT:    # kill: def $eax killed $eax killed $rax
+; CHECK-NEXT:    retq
+entry:
+  %asm_call = call i32 asm sideeffect "", "=r,0,~{dirflag},~{fpsr},~{flags}"(ptr %f)
```
Is it really allowed to have the mismatched sizes, pointer 64 with i32? Should that be an x86-only thing?

arsenm replied:
What about double + i32? Or i16 + float?

dfszabo commented:
Well if approved, then can someone merge it?

arsenm committed:
Merge branch 'main' into main-inline-asm-fix-fp-int-move
